compat/src/test/csharp/Program.cs

```diff
@@ -0,0 +1,399 @@
+        "varchar" => NpgsqlDbType.Varchar,
+        "timestamp" => NpgsqlDbType.Timestamp,
+        "date" => NpgsqlDbType.Date,
+        "char" => NpgsqlDbType.Char,
+        _ => throw new ArgumentException($"Unsupported type: {type}")
+    };
+
+    private object ConvertParameterValue(object value, string type)
+    {   
+        return type.ToLower() switch
+        {
+            "int4" or "int8" => Convert.ToInt64(value, CultureInfo.InvariantCulture),
+            "float4" or "float8" => Convert.ToDouble(value.ToString(), CultureInfo.InvariantCulture),
+            "boolean" => Convert.ToBoolean(value),
+            "varchar" => value.ToString()!,
+            "date" => DateTime.Parse(value.ToString()!, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind),
+            "char" => Convert.ToChar(value),
+            "timestamp" => DateTime.Parse(value.ToString()!, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind),
+            _ => value
+        };
+    }
+
+    private void AssertResult(ExpectDefinition expect, object actual)
+    {
+        var actualConverted = ConvertQueryResult(actual);
+
+        if (expect.Result != null)
+        {
+            if (expect.Result is List<object> expectedResult)
+            {
+                if (!DeepEquals(actualConverted, expectedResult))
+                {
+                    throw new AssertionException(
+                        $"Expected result {System.Text.Json.JsonSerializer.Serialize(expectedResult)}, got {System.Text.Json.JsonSerializer.Serialize(actualConverted)}");
+                }
+            }
+            else
+            {
+                // Single value comparison
+                if (!actual.ToString()!.Equals(expect.Result.ToString()))
+                {
+                    throw new AssertionException($"Expected result '{expect.Result}', got '{actual}'");
+                }
+            }
+        }
+        else if (expect.ResultContains != null)
+        {
+            foreach (var expectedRow in expect.ResultContains)
+            {
+                var expectedRowList = expectedRow as List<object>;
+                if (expectedRowList == null)
+                {
+                    throw new AssertionException($"Expected row {expectedRow} is not in the correct format");
+                }
+
+                if (!actualConverted.Any(row => DeepEquals(row, expectedRowList)))
+                {
+                    throw new AssertionException(
+                        $"Expected row {System.Text.Json.JsonSerializer.Serialize(expectedRow)} not found in actual results {System.Text.Json.JsonSerializer.Serialize(actualConverted)}");
+                }
+            }
+        }
+    }
+
+    private List<List<object>> ConvertQueryResult(object result)
+    {
+        if (result is not List<Dictionary<string, object>> dictList)
+            return new List<List<object>>();
+
+        return dictList.Select(dict => dict.Values
```

Comment:
Just for my interest, `dict.Values` seems to be iterated over in pseudo-random order? Do you then have an assertion that expects this exact order?

Reply:
yeah, assertions do require the exact order.
I changed it to use OrdereredDictionary which I understood guarantees insertion order.